lib/reply.js

```diff
@@ -202,48 +202,22 @@ function sendStream (payload, res, reply) {
 }
 
 function handleError (reply, error, cb) {
-  var statusCode = reply.res.statusCode
-  if (error == null) {
-    statusCode = statusCode || 500
-  } else if (error.status >= 400) {
-    if (error.status === 404) {
-      notFound(reply)
-      return
-    }
-    statusCode = error.status
-  } else if (error.statusCode >= 400) {
-    if (error.statusCode === 404) {
-      notFound(reply)
-      return
-    }
-    statusCode = error.statusCode
-  } else {
-    statusCode = statusCode || 500
+  var statusCode = getErrorStatusCode(reply.res.statusCode, error)
+  if (statusCode === 404 && reply.res.statusCode !== 404) {
```

Comment:
This doesn't seem like a good idea. It's possible for a user to set the status code to `404` and then send a 404 error after. If the error object has a `404` status then the not-found handler should be called regardless of the value of `res.statusCode`.

Reply:
If that's the case, there's a bug with current code.
Because the unit tests failed when the following code was added during refactor:
```
if (statusCode === 404) {
  notFound(reply)
  return
}
```

Reply:
@nwoltman I can create a separate issue to explain the above problem, so that it can be taken up in different PR.
I'm new to fastify code base, and I don't want to modify unit tests as part of this refactor.

Reply:
There isn't a bug with the current code. The not-found handler should only be called if the error object has a `404` status. If `res.statusCode` is `404` but the error's status is not `404`, the not-found handler should not be called. Your refactored code handles this properly, except for the case when both `res.statusCode` and `error.status` are `404` (which there currently is no test for).

Reply:
What should be the behavior when both `res.statusCode` and `error.status`/`error.statusCode` are `404`?

Reply:
I think the status code in the error should prevail. I do not think this is specified atm.

Reply:
I've created #753 to discuss this issue